File: jdevkit/src/main/java/com/github/klee0kai/android_devkit/threads/DefaultThreadFactory.java
package com.github.klee0kai.android_devkit.threads;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;

public class DefaultThreadFactory implements ThreadFactory {

    private static final AtomicInteger poolNumber = new AtomicInteger(1);
    private final ThreadGroup group;
    private final AtomicInteger threadNumber = new AtomicInteger(1);
    private final String namePrefix;
    private final boolean isDaemonThreads;

    public DefaultThreadFactory(String poolname, boolean isDaemonThreads) {
        this.isDaemonThreads = isDaemonThreads;
        SecurityManager s = System.getSecurityManager();
        group = (s != null) ? s.getThreadGroup() :
                Thread.currentThread().getThreadGroup();
        namePrefix = poolname + "-" +
                poolNumber.getAndIncrement() +
                "-thread-";
    }

    @Override
    public Thread newThread(Runnable r) {
        Thread t = new Thread(group, r, namePrefix + threadNumber.getAndIncrement(), 0);
        if (t.isDaemon())
            t.setDaemon(isDaemonThreads);
        if (t.getPriority() != Thread.NORM_PRIORITY)
            t.setPriority(Thread.NORM_PRIORITY);
        return t;
    }

}

File: kotlin_hummus/src/main/kotlin/com/github/klee0kai/hummus/threads/DefaultThreadFactory.kt
package com.github.klee0kai.hummus.threads

import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger

open class DefaultThreadFactory(
    poolname: String,
    private val isDaemonThreads: Boolean
) : ThreadFactory {

    private val group: ThreadGroup
    private val threadNumber = AtomicInteger(1)
    private val namePrefix: String

    init {
        val s = System.getSecurityManager()
        group = if (s != null) s.threadGroup else Thread.currentThread().threadGroup as ThreadGroup
        namePrefix = poolname + "-" + poolNumber.getAndIncrement() + "-thread-"
    }

    override fun newThread(r: Runnable): Thread {
        val t = Thread(group, r, namePrefix + threadNumber.getAndIncrement(), 0)
        if (t.isDaemon) t.isDaemon = isDaemonThreads
        if (t.priority != Thread.NORM_PRIORITY) t.priority = Thread.NORM_PRIORITY
        return t
    }

    companion object {
        private val poolNumber = AtomicInteger(1)
    }
}